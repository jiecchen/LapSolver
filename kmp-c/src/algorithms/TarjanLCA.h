#pragma once

#include "structures/Graph.h"
#include "structures/TreeChildren.h"

struct TarjanLCA {
    TarjanLCA (const Graph& g, int tRoot, const int* tParent, const TreeChildren& tChildren, int* dfsOrder);
    ~TarjanLCA ();

    int ne;
    int* u;
    int* v;
    double* weight;
    int* lca;
};