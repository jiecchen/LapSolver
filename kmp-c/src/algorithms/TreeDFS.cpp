#include <vector>
#include "structures/TreeChildren.h"

using namespace std;

vector<int> TreeDFS(int nv, int root, const TreeChildren& children) {
    int* dfsStack = new int[nv]; dfsStack[0] = root;
    int dfsStackPos = 1;
    vector<int> dfsOrder;
    dfsOrder.reserve(nv);

    while (dfsStackPos > 0) {
        int curNode = dfsStack[--dfsStackPos];
        dfsOrder.push_back(curNode);
        for (int i = children.offset[curNode]; i < children.offset[curNode+1]; i++) {
            dfsStack[dfsStackPos++] = children.child[i];
        }
    }

    delete[] dfsStack;

    return dfsOrder;
}
