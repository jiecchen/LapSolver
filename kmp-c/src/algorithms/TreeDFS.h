#pragma once

#include <vector>
#include "structures/TreeChildren.h"

std::vector<int> TreeDFS(int nv, int root, const TreeChildren& children);