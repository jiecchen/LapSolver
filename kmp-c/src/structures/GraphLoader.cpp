#include "GraphLoader.h"
#include <fstream>
#include <sstream>
#include <string>
using namespace std;

Graph GraphLoader::fromArrays(aligned_vector<int> u, aligned_vector<int> v, aligned_vector<double> w)
{
    return Graph(EdgeList(u, v, w));
}

Graph GraphLoader::fromFile(const string &filename)
{
    aligned_vector<int> u = aligned_vector<int>();
    aligned_vector<int> v = aligned_vector<int>();
    aligned_vector<double> w = aligned_vector<double>();

    string line;
    ifstream ijvFile(filename);
    if (!ijvFile)
        throw -1;

    while (getline(ijvFile, line))
    {
        if (line.empty()) break;
        istringstream iss(line);
        int a, b;
        double c;
        if (!(iss >> a >> b >> c))
            throw u.size();
        u.push_back(a);
        v.push_back(b);
        w.push_back(c);
    }

    return fromArrays(u, v, w);
}