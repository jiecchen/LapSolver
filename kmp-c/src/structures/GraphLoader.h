#pragma once

#include <string>
#include <iostream>
#include "structures/Graph.h"
#include "util/aligned.h"

class GraphLoader
{
public:
	
	static Graph fromArrays(aligned_vector<int> u, aligned_vector<int> v, aligned_vector<double> w);
	static Graph fromFile(const std::string &filename);
    static Graph fromStream(std::istream &inStream);
    static Graph fromStdin();

};