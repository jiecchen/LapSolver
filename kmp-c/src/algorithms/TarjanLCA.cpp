#include <iostream>

#include "structures/Graph.h"
#include "structures/TreeChildren.h"
#include "TarjanLCA.h"
#include "UnionFind.h"
#include "TreeDFS.h"

using namespace std;

TarjanLCA::TarjanLCA (const Graph& g, int tRoot, const int* tParent, const TreeChildren& tChildren, int* dfsOrder) {
    int n = g.nv();

    // run LCA
    int* ancestor = new int[n];
    bool* black = new bool[n];
    int* childrenLeft = new int[n];

    UnionFind components(n);

    int edgePos = 0;
    ne = g.ne() - n + 1;
    u = new int[ne];
    v = new int[ne];
    lca = new int[ne];

    black[0:n] = false; 
    childrenLeft[0:n] = tChildren.offset[1:n] - tChildren.offset[0:n];
    for (int i = 0; i < n; i++) ancestor[i] = i;

    for (int i = n-1; i >= 0; i--) {
        int curNode = dfsOrder[i];
        int parent = tParent[curNode];

        if (curNode != tRoot) {
            childrenLeft[parent]--;
        }

        // done with all children
        if (!childrenLeft[curNode]) {
            black[curNode] = true;

            for (int i = 0; i < g.degree(curNode); i++) {
                int queryNode = g.neighbor(curNode, i);
                if (black[queryNode] && tParent[queryNode] != curNode && tParent[curNode] != queryNode) {
                    u[edgePos] = curNode;
                    v[edgePos] = queryNode;
                    lca[edgePos] = ancestor[components.find_set(queryNode)];
                    edgePos++;
                }
            }

            components.link(parent, curNode);
            ancestor[components.find_set(parent)] = parent;
        }
    }

    delete[] ancestor;
    delete[] black;
    delete[] childrenLeft;
}

TarjanLCA::~TarjanLCA () {
    delete[] u;
    delete[] v;
    delete[] lca;
}
