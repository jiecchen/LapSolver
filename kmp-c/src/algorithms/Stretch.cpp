#include "structures/Graph.h"
#include "structures/TreeChildren.h"
#include "TarjanLCA.h"
#include "Stretch.h"

Stretch::Stretch (const Graph& g, int tRoot, const int* tParentIndex, int* dfsOrder, const TarjanLCA& lca) {
    int n = g.nv();

    // compute resistances to root top-down
    double* resistanceToRoot = new double[n];
    resistanceToRoot[tRoot] = 0;
    for (int i = 1; i < n; i++) {
        int u = dfsOrder[i];
        int parent = g.neighbor(u, tParentIndex[u]);
        resistanceToRoot[u] = resistanceToRoot[parent] + 1/g.weight(u, tParentIndex[u]);
    }

    // get stretches
    total = 0;
    stretch = new double[lca.ne];
    for (int i = 0; i < lca.ne; i++) {
        stretch[i] = resistanceToRoot[lca.u[i]] + resistanceToRoot[lca.v[i]] - 2*resistanceToRoot[lca.lca[i]];
        total += stretch[i];
    }
}

Stretch::~Stretch () {
    delete[] stretch;
}
