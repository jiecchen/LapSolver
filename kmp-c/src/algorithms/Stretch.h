#pragma once

#include "structures/Graph.h"
#include "structures/TreeChildren.h"
#include "TarjanLCA.h"

struct Stretch {
    Stretch (const Graph& g, int tRoot, const int* tParentIndex, int* dfsOrder, const TarjanLCA& lca);
    ~Stretch ();

    double total;
    double* stretch;
};