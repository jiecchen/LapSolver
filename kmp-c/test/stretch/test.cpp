// Test for Tarjan LCA

#include <cstdlib>
#include <iostream>
#include <string>
#include <algorithm>
#include <algorithms/TarjanLCA.h>
#include <algorithms/TreeDFS.h>
#include <algorithms/Stretch.h>
#include <structures/TreeChildren.h>
#include <structures/GraphLoader.h>

using namespace std;

int main(int argc, char *argv[])
{
    atexit(mkl_free_buffers);

    // read graph
    Graph g = GraphLoader::fromStdin();

    // read spanning tree
    int n = g.nv();
    vector<int> parent(n);
    int root;
    for (int i = 0; i < n; i++) {
        cin >> parent[i];
        if (parent[i] == i) root = i;
    }

    // precompute children and dfs order
    TreeChildren tChildren(n, parent.data());
    vector<int> dfsOrder = TreeDFS(n, root, tChildren);

    // crudely get parent indices
    vector<int> parentIndex(n);
    for (int u = 0; u < n; u++) {
        if (u == root) continue;
        for (int i = 0; i < g.degree(u); i++) {
            if(g.neighbor(u, i) == parent[u]) parentIndex[u] = i;
        }
    }

    // run lca then stretch
    TarjanLCA lca(g, root, parent.data(), tChildren, dfsOrder.data());
    Stretch st(g, root, parentIndex.data(), dfsOrder.data(), lca);

    // canonize lca results
    for (int i = 0; i < lca.ne; i++) {
        if (lca.u[i] > lca.v[i]) swap(lca.u[i], lca.v[i]);
    }

    vector<int> ord(lca.ne);
    for (int i = 0; i < lca.ne; i++) ord[i] = i;
    sort(ord.begin(), ord.end(), [&] (int a, int b) {
        if (lca.u[a] != lca.u[b]) return lca.u[a] < lca.u[b];
        return lca.v[a] < lca.v[b];
    });

    // read off sorted lca results
    for (int i = 0; i < lca.ne; i++) {
        printf("stretch(%d, %d) = %.6lf\n", lca.u[ord[i]], lca.v[ord[i]], st.stretch[ord[i]]);
    }

    return 0;
}
