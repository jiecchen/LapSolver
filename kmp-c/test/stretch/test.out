stretch(0, 3) = 2.000000
stretch(0, 4) = 2.000000
stretch(0, 5) = 2.000000
stretch(0, 6) = 2.000000
stretch(1, 2) = 2.000000
stretch(1, 5) = 3.000000
stretch(1, 6) = 3.000000
stretch(2, 3) = 3.000000
stretch(2, 4) = 3.000000
stretch(3, 4) = 2.000000
stretch(3, 5) = 4.000000
stretch(3, 6) = 4.000000
stretch(4, 5) = 4.000000
stretch(4, 6) = 4.000000
stretch(5, 6) = 2.000000
